### ISSUE TITLE
<format>: fuzz the format string parser

### ISSUE THREAD

**Issue by barcharcraz** [microsoft/STL#1946]
It would be nice to fuzz the format string parser, although it's not designed to parse untrusted data (format strings should really be literals), it would be nice to have some piece of mind. Also fuzzing could help make our error messages better.

It's not completely clear how this should work, what fuzzer to use, how to integrated it into the test harness, and if we should try and make it apply to llvm's format implementation as well.

**Comment by pauldreik** [comment]
If it is not possible to run the actual fuzzing, it would definitely be possible to let the fuzzer corpus from fmt on oss fuzz be replayed on a sanitized msvc build. That would definitely be better than not testing at all. The fuzz corpus is several MB with thousands of weird inputs.

**Comment by pauldreik** [comment]
I made a limited effort for a few hours and just wanted to write down what I did and what happened.

TLDR: replaying fuzz data for the one arg fuzzer through a msvc address sanitizer build revealed no errors (for the format version included in VS2019 16.11.2). Making a clang fuzz build on windows is perhaps possible, but won't succeed because msvc format needs concepts, which clang does not support (at least by default on windows, as of now).

I used Visual Studio 2019 (16.11.2)

I took the fmt project and built it with msvc, enabling address sanitizer and c++latest manually in the visual studio GUI. The fmt fuzzers already have a [replay mode by default](https://github.com/fmtlib/fmt/blob/74c1118964d0d943535c03d6c4e6f2d1449b61d6/test/fuzzing/main.cc#L9-L20), which I had to slightly modify (opening the files in binary mode to avoid newline conversion). I inserted a call to std::format(...) next to the fmt::format(...) in the fuzzer.

At this point, the fuzz data (downloaded from oss-fuzz, took the latest available corpus backup as of today 20210828) passed through std::format.

Because weird input causes huge allocations inside std::format, grinding it to a halt, I had to modify <format>.  Nothing strange with that, otherwise it takes tens of seconds or even more per fuzz case. I did that by copying it locally into "msvc_format.h" and modifying _Grow with:

```cpp
 void _Grow(size_t _Capacity) final {
    if (_Capacity > 123000) {
      throw std::runtime_error("fuzz too much");
    }
    _Cont.resize(_Capacity);
    this->_Set(_RANGES data(_Cont), _Capacity);
  }
```
I did a similar thing at another location in format which I suspect might not be needed:
```cpp
_NODISCARD inline string vformat(const string_view _Fmt,
                                 const format_args _Args) {
  string _Str;
  auto requested = _Fmt.size() + _Args._Estimate_required_capacity();
  if (requested > 1000000) {
    throw std::runtime_error("fuzz - too much");
  }
```

Now the fuzz corpus passed through without problems. I tried both debug and release mode. I did not go further with the [chrono fuzzer](https://github.com/fmtlib/fmt/blob/master/test/fuzzing/chrono-duration.cc) for now, I only used the [onearg fuzzer from fmt.](https://github.com/fmtlib/fmt/blob/master/test/fuzzing/one-arg.cc)

Next step was to see if I could make a clang build (using clang from the visual studio installer) using libfuzzer. I was able to make partial progress and create a fuzz instrumented binary. I stopped because std::format on msvc is [not usable without concepts](https://github.com/microsoft/STL/blob/78ff4618d4fa43eeb4bb714d76f27d88cf11a1b6/stl/inc/format#L43). So this work could be revisited once clang on windows has concepts. Or is there a way to disable the need for concepts in the format header?

**Comment by CaseyCarter** [comment]
> I made a limited effort for a few hours and just wanted to write down what I did and what happened.
> 
> TLDR: replaying fuzz data for the one arg fuzzer through a msvc address sanitizer build revealed no errors (for the format version included in VS2019 16.11.2). Making a clang fuzz build on windows is perhaps possible, but won't succeed because msvc format needs concepts, which clang does not support (at least by default on windows, as of now).

Clang has supported Concepts on Windows for as long as clang has supported Concepts: since version 10. We support clang as a first-class consumer of the STL, and run all of our tests (well, all the tests that aren't specifically for managed things) with Clang in every CI run. Specifying C++20 should be enough to get you Concepts on a contemporary Clang. (One selects C++20 either via `-std=c++20` for `clang`, or `/std:c++latest` for `clang-cl` which won't grok `/std:c++20` until the version 13 release.)

**Comment by pauldreik** [comment]
Thanks @CaseyCarter but it does not seem to work even when specifying C++20 or latest.

This is the program:
```
C:\Users\slask\Documents\code\fmt>type format.cpp
#include <format>
#include <iostream>
int main(int argc, char* argv[]) {
        std::cout<<std::format("hello {}","world")<<std::endl;
}
```

Compilation with cl, works fine as expected:
```
C:\Users\slask\Documents\code\fmt>cl  /std:c++latest format.cpp
Microsoft (R) C/C++ Optimizing Compiler Version 19.29.30133 for x86
Copyright (C) Microsoft Corporation.  All rights reserved.

/std:c++latest is provided as a preview of language features from the latest C++
working draft, and we're eager to hear about bugs and suggestions for improvements.
However, note that these features are provided as-is without support, and subject
to changes or removal as the working draft evolves. See
https://go.microsoft.com/fwlink/?linkid=2045807 for details.

format.cpp
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.29.30133\include\ostream(515): warning C4530: C++ exception handler used, but unwind semantics are not enabled. Specify /EHsc
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.29.30133\include\ostream(508): note: while compiling class template member function 'std::basic_ostream<char,std::char_traits<char>> &std::basic_ostream<char,std::char_traits<char>>::put(_Elem)'
        with
        [
            _Elem=char
        ]
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.29.30133\include\ostream(979): note: see reference to function template instantiation 'std::basic_ostream<char,std::char_traits<char>> &std::basic_ostream<char,std::char_traits<char>>::put(_Elem)' being compiled
        with
        [
            _Elem=char
        ]
format.cpp(4): note: see reference to class template instantiation 'std::basic_ostream<char,std::char_traits<char>>' being compiled
Microsoft (R) Incremental Linker Version 14.29.30133.0
Copyright (C) Microsoft Corporation.  All rights reserved.

/out:format.exe
format.obj
```
(the output is format.exe which prints the expected output).

If I use clang (12.0.0 from the VS installation):
```
C:\Users\slask\Documents\code\fmt>"C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\Llvm\bin\clang-cl.exe"  /std:c++latest format.cpp
format.cpp(4,18): error: no member named 'format' in namespace 'std'
        std::cout<<std::format("hello {}","world")<<std::endl;
                   ~~~~~^
1 error generated.
```

**Comment by StephanTLavavej** [comment]
@pauldreik It's a long story, but for Clang 12 you need to compile with `/std:c++latest /D_HAS_CXX23` in order to activate `std::format` (which we're currently guarding in C++23 mode, until we're done implementing all C++20 Defect Reports). For example, see:

https://github.com/microsoft/STL/blob/d6f9987d7ec694b8599fe21e87e3653c36566223/tests/std/tests/usual_matrix.lst#L40

**Comment by pauldreik** [comment]
I revisited this now, things have really gone forward! I used a visual studio 2022 command prompt and built my chrono fuzzer [available here](https://github.com/pauldreik/stdfuzz) using

> clang++ -fsanitize=fuzzer,undefined -std=c++2b format\chrono-duration.cpp -I helpers 

it is fuzzing now, will check tomorrow. Most errors are usually found pretty quick, so I do not expect it to find anything.
